### PR TITLE
Add Fomu support

### PR DESCRIPTION
**Pull request by piotr-binkowski**
The aim of this PR is to add support for Fomu.
Initial design uses the Foboot configuration/firmware which requires special handling because it needs a custom BIOS, due to that at this point it is necessary to first run `make foboot-bios` to build the replacement BIOS before building other targets like `gateware`
It is also required to manually patch foboot with `firmware/foboot/putchar-fix.patch`

### CODE REVIEW

File: rtl/2-stage-1024-cache.v_toplevel_RegFilePlugin_regFile.bin [hunk 1]
```diff
@@ -0,0 +1,32 @@
+00000000000000000000000000000000
```
Comment (mithro) [inline rtl/2-stage-1024-cache.v_toplevel_RegFilePlugin_regFile.bin]: This CL shouldn't be adding an `rtl` directory. All `rtl` is under the `gateware` directory.

File: targets/fomu/Makefile.mk [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+BAUD ?= 115200
+
+MAKE_LITEX_EXTRA_CMDLINE += "--no-compile-software"
+
+FOBOOT_BIN = $(TARGET_BUILD_DIR)/../../rtl/foboot.bin
+FOBOOT_DIR = $(TARGET_BUILD_DIR)/../../third_party/foboot/sw
+
+# Image
+image-flash-$(PLATFORM):
+	iceprog $(IMAGE_FILE)
```
Comment (mithro) [inline targets/fomu/Makefile.mk]: `iceprog` isn't use with Fomu -- you want `dfu-util`

File: rtl/2-stage-1024-cache.v_toplevel_RegFilePlugin_regFile.bin (hunk 1 above)
Comment (piotr-binkowski) [reply rtl/2-stage-1024-cache.v_toplevel_RegFilePlugin_regFile.bin]: I've moved files from `rtl` directory to `gateware`

Comment (piotr-binkowski) [reply rtl/2-stage-1024-cache.v_toplevel_RegFilePlugin_regFile.bin]: I've moved files from rtl directory to gateware

File: targets/fomu/Makefile.mk (hunk 2 above)
Comment (piotr-binkowski) [reply targets/fomu/Makefile.mk]: At this point the `base` target for Fomu builds the whole bootloader image (gateware + foboot as bios). I think that loading it must be done using `iceprog` and a direct flash connection. Or is it possible to update the bootloader using `dfu-util`?

**Comment by piotr-binkowski** [comment]
@mithro please take a look
I think that I addressed all of your comments and if everything is all right then this can be merged

File: targets/fomu/Makefile.mk (hunk 2 above)
Comment (mithro) [reply targets/fomu/Makefile.mk]: It is possible to update the bootloader via dfu-util

File: targets/fomu/base.py [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+        self.comb += self.cd_usb_12.clk.eq(clk12)
+
+        self.sync.por += \
+            If(reset_delay != 0,
+                reset_delay.eq(reset_delay - 1)
+            )
+        self.specials += AsyncResetSynchronizer(self.cd_por, self.reset)
+
+
+class SBLED(Module, AutoCSR):
```
Comment (mithro) [inline targets/fomu/base.py]: This should be in `gateware` directory. I believe there is a version at https://github.com/timvideos/litex-buildenv/blob/master/gateware/ice40.py

File: targets/fomu/base.py [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+            # o_LEDDON = led_is_on, # Indicates whether LED is on or not
+            # i_LEDDRST = ResetSignal(), # This port doesn't actually exist
+            o_PWMOUT0 = ledd_value[0],
+            o_PWMOUT1 = ledd_value[1],
+            o_PWMOUT2 = ledd_value[2],
+            o_LEDDON = Signal(),
+        )
+
+
+class SBWarmBoot(Module, AutoCSR):
```
Comment (mithro) [inline targets/fomu/base.py]: Should be in https://github.com/timvideos/litex-buildenv/blob/master/gateware/ice40.py

File: targets/fomu/base.py [hunk 5]
```diff
@@ -0,0 +1,451 @@
...
+            touch1.oe.eq(self.oe.storage[0]),
+            touch2.oe.eq(self.oe.storage[1]),
+            touch3.oe.eq(self.oe.storage[2]),
+            touch4.oe.eq(self.oe.storage[3]),
+
+            self.i.status.eq(Cat(touch1.i, touch2.i, touch3.i, touch4.i))
+        ]
+
+
+class PicoRVSpi(Module, AutoCSR):
```
Comment (mithro) [inline targets/fomu/base.py]: Should be in `gateware`.

**Comment by mithro** [comment]
The aim here is to have a SoC you can load into the "user" part of the Fomu. This is the image that boots *after* the bootloader. See the "Migen and LiteX" section of https://workshop.fomu.im/#migen-and-litex

**Comment by mithro** [comment]
Could you please rebase this pull request?

**Comment by mithro** [comment]
@mateusz-holenko -- IIRC Could you work with @piotr-binkowski and @ewenmcneill to get this pull request merged?

**Comment by piotr-binkowski** [comment]
I've rebased this and moved module definitions to `gateware` but I still need to update this so it generates a proper `user` image and not a bootloader.

**Comment by piotr-binkowski** [comment]
Target now generates a normal user image that works without BIOS (assigning constant values to LEDs in gateware shows proper colors) so only thing missing now is booting BIOS or other software from Flash via XIP since BIOS doesn't fit inside BRAMs.

**Comment by piotr-binkowski** [comment]
I've updated the target to use BIOS stored at 0x60000 in Flash (user gateware base address + gateware size) and I've modified the platform to by default target PVT variant. During testing I also noticed an issue with yosys version that is currently used but I will create an separate issue for that.
Please take a look @mithro

**Pull request merged**